File: Matrix/Net.cpp
#include "Matrix.h"

Matrix sigmoid(Matrix m){
	m.forEach([](double a){return 1.0 / (1.0 + exp(-a));});
	return m;
}
Matrix sigmoidPrime(Matrix m){
	m.forEach([](double a){
			auto s = 1.0/(1.0 + exp(-a));
			return s*(1.0-s);
			});
	return m;
}
class Layer{
	friend class Net;
	private:
		int l;
		Matrix I,O,G;
	public:
		Layer(int l):l(l){}
};

class Net{
	private:
		std::vector<Layer> L;
		std::vector<Matrix> W;
	public:
		Net(std::vector<int> topology);
		Matrix FF(Matrix& X);
		Matrix BP(Matrix& X, Matrix& Y);
};


Net::Net(std::vector<int> T){
	for(size_t i=0; i<T.size(); ++i){
		L.push_back(Layer(T[i]));
	}
	for(size_t i=1;i<T.size(); ++i){
		W.push_back(Matrix(T[i],T[i-1]));
	}

}

Matrix Net::FF(Matrix& X){
	L[0].O = X;
	for(size_t i=1; i<L.size(); ++i){
		L[i].I = dot(W[i-1],L[i-1].O);
		L[i].O = sigmoid(L[i].I);
	}
	return L.back().O;
}

Matrix Net::BP(Matrix& X, Matrix& T){
	auto Y = FF(X);
	auto G = T - Y;
	L.back().G = G * sigmoidPrime(L.back().O);
	for(int i = L.size()-1; i>=1; --i){
		L[i].G = dot(L[i+1].G,W[i].transpose()) * sigmoidPrime(L[i].I);
	}
}

File: NeuralNetwork/Net.py
import numpy as np
from matplotlib import pyplot as plt


def sigmoid(x):
    return 1.0/(1.0 + np.exp(-x))
def sigmoidPrime(x):
    s = sigmoid(x)
    return s * (1.0 - s)

class Layer:
    def __init__(self,l):
        pass


class Net:
    def __init__(self,t):
        self.length = len(t)
        self.L = []
        self.W = [] 
        for l in t:
            self.L.append(Layer(l))
        for i in range(1, self.length):
            self.W.append(np.random.randn(t[i-1],t[i]))
    def FF(self,X):
        self.L[0].O = X
        print(X)
        for i in range(1,self.length):
            self.L[i].I = np.dot(self.L[i-1].O,self.W[i-1])
            self.L[i].O = sigmoid(self.L[i].I)
        return self.L[-1].O
    def BP(self,X,Y):
        #currently writing here, incomplete.
        self.L[-1].G = (self.FF(X)-Y) * sigmoidPrime(self.L[i-1].O)
        for i in range(self.length-1, 1):
            L[i].G = L[i+1].G * sigmoidPrime(self.L[i].I)
        for i in range(1, self.length):
            dW = np.dot(L[i].G, self.L[i]
        pass

def main():
    t = [2,4,1]
    net = Net(t)
    print net.FF(np.atleast_2d([0,0]))
if __name__ == "__main__":
    main()
